### PR TITLE
Refactor Riichi point calculations

### PR DESCRIPTION
**Pull request by TwelveNights**
Resolution for #16 and #8

### CODE REVIEW

File: imports/api/Constants.js [hunk 1]
```diff
@@ -34,7 +34,14 @@ export const Constants = {
...
 	},
-	MAHJONG_CLUB_LEAGUE: "Mahjong Club League"
+
+	EAST: "east",
+	SOUTH: "south",
+	WEST: "west",
+	NORTH: "north",
+
+	MAHJONG_CLUB_LEAGUE: "Mahjong Club League",
+	MANGAN: 2000
```
Comment (Futochu) [inline imports/api/Constants.js]: You have good intentions here, but "MANGAN = 2000" is meaningless.  You may as well have called the constant "Y2K" or "Andre minus 1000".  At least call this something like MANGAN_BASE_POINTS or something.

File: imports/api/NewGameUtils.js [hunk 2]
```diff
@@ -47,60 +47,58 @@ export var NewGameUtils = {
...
-			break;
-		case Constants.GAME_TYPE.JAPANESE:
-			if (round <= 4)
-				return "東";
-			if (round > 4 && round <= 8)
-				return "南";
-			else //if (round > 8)
-				return "西";;
-			break;
+			case Constants.GAME_TYPE.HONG_KONG:
```
Comment (Futochu) [inline imports/api/NewGameUtils.js]: This is not the style I use for switch/case.  Case stays indented the same as switch.

File: imports/api/NewGameUtils.js [hunk 3]
```diff
@@ -47,60 +47,58 @@ export var NewGameUtils = {
...
-		case Constants.GAME_TYPE.HONG_KONG:
-			if (round <= 4)
-				return "東";
-			if (round > 4 && round <= 8)
-				return "南";
-			if (round > 8 && round <= 12)
-				return "西";
-			else //if (round > 12)
-				return "北";
-			break;
```
Comment (Futochu) [inline imports/api/NewGameUtils.js]: Why are you so keen on putting breaks in every case?  I prefer to not include breaks in switch cases which are returned from.  Looks cleaner and the break is never called anyway.

File: imports/api/NewGameUtils.js [hunk 4]
```diff
@@ -47,60 +47,58 @@ export var NewGameUtils = {
...
+				if (round > 4 && round <= 8)
+					return "南";
+				else //if (round > 8)
+					return "西";
 		};
 	},
 
 	// Helper function to ensure all players are selected
 	allPlayersSelected() {
-		return (Session.get("current_east") != Constants.DEFAULT_EAST &&
```
Comment (Futochu) [inline imports/api/NewGameUtils.js]: This is just the style I've been taught, but can formatting please be put in a seperate branch if you think it's important?  Otherwise there's just a lot of clutter.

Comment (Futochu) [inline imports/api/NewGameUtils.js]: Most of this pull request seems unrelated to point calculation.  Either rename the commit or change this to reflect what the commit is.

File: imports/api/PointCalculationUtils.js [hunk 5]
```diff
@@ -0,0 +1,118 @@
+import { Constants } from './Constants';
+import { NewGameUtils } from './NewGameUtils';
+
+export var PointCalculationUtils = {
```
Comment (Futochu) [inline imports/api/PointCalculationUtils.js]: If you're going to do this, make it generic (include HK calculation too it's much more simple anyway)

File: imports/ui/JapaneseNewGame.js [hunk 6]
```diff
@@ -601,7 +601,7 @@ function save_game_to_database(hands_array) {
 	};
 
 	// Initialise ELO calculator to update player ELO
-	var jpn_elo_calculator = new EloCalculator(2000, 5, [15000, 0, -5000, -10000], game, Constants.GAME_TYPE.JAPANESE);
+	var jpn_elo_calculator =  new EloCalculator(2000, 5, [15000, 0, -5000, -10000], game, Constants.GAME_TYPE.JAPANESE);
```
Comment (Futochu) [inline imports/ui/JapaneseNewGame.js]: Remove this space

File: imports/ui/JapaneseNewGame.js [hunk 7]
```diff
@@ -720,42 +720,44 @@ function push_dealin_hand(template) {
...
 		Session.set("eastPlayerWins", Number(Session.get("eastPlayerWins")) + 1);
 		Session.set("eastPlayerPointsWon", Number(Session.get("eastPlayerPointsWon")) + points);
 		Session.set("eastPlayerDoraSum", Number(Session.get("eastPlayerDoraSum")) + dora);
-		if (Session.get("east_riichi") == true) 
+		if (Session.get("east_riichi") == true) {
 			Session.set("eastPlayerRiichisWon", Number(Session.get("eastPlayerRiichisWon")) + 1);
-	}
-	else if (winnerWind == "south") {
+		}
+	} else if (winnerWind == Constants.SOUTH) {
```
Comment (Futochu) [inline imports/ui/JapaneseNewGame.js]: Why are you so dead-set on changing all my style to your preferred style?  Leave brackets and indentation the way they are.

File: imports/api/PointCalculationUtils.js [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+        "selfdraw_delta": selfdraw_delta,
+        "fuckup_delta": fuckup_delta
+    }
+};
+
+function dealin_delta(points, fu, winnerWind, loserWind, riichiSticks) {
+    let winds = {};
+    winds[Constants.EAST] = winds[Constants.SOUTH] = winds[Constants.WEST] = winds[Constants.NORTH] = 0;
+
+	let basicPoints;
```
Comment (Futochu) [inline imports/api/PointCalculationUtils.js]: Some funky indentation here

File: imports/api/PointCalculationUtils.js [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+    let multiplier = (winnerWind != NewGameUtils.roundToDealerDirection(Number(Session.get("current_round")))) ? 4 : 6;
+    let manganPayout = multiplier * Constants.MANGAN;
+
+	if (points < 5) {
+        if (fu == 20 || (points == 1 && fu == 25)) {
```
Comment (Futochu) [inline imports/api/PointCalculationUtils.js]: Super funky indentation.  Use tabs for indentation and spaces for alignment please.

File: imports/api/NewGameUtils.js (hunk 3 above)
Comment (TwelveNights) [inline imports/api/NewGameUtils.js]: It wasn't me that had them to begin with lmao

File: imports/api/Constants.js (hunk 1 above)
Comment (TwelveNights) [inline imports/api/Constants.js]: Sure

File: imports/api/NewGameUtils.js (hunk 2 above)
Comment (TwelveNights) [inline imports/api/NewGameUtils.js]: I'll update my code to be like that

File: imports/api/PointCalculationUtils.js (hunk 5 above)
Comment (TwelveNights) [inline imports/api/PointCalculationUtils.js]: I'd like to do it later, since this ticket is already quite large.

File: imports/api/PointCalculationUtils.js (hunk 8 above)
Comment (TwelveNights) [inline imports/api/PointCalculationUtils.js]: I'll fix that in my next commit.

File: imports/api/NewGameUtils.js (hunk 3 above)
Comment (Futochu) [inline imports/api/NewGameUtils.js]: o im dum

File: imports/ui/JapaneseNewGame.js (hunk 7 above)
Comment (TwelveNights) [inline imports/ui/JapaneseNewGame.js]: Force of habit, although keep in mind that all this change does is modify it so that the oneline if statement is encapsulated in curly braces instead of following in a logical line.

Comment (Futochu) [inline imports/ui/JapaneseNewGame.js]: Just an encompasing statement I happened to put on this line. I actually generally agree with wrapping single line conditional logic.

**Comment by Futochu** [comment]
Some responses posted.  Please refactor as well as there are now merge conflicts.

File: imports/api/Constants.js [hunk 10]
```diff
@@ -43,7 +46,21 @@ export const Constants = {
...
+
+	PRIORITY: {
+		east: 3,
+		south: 2,
+		north: 1,
+		west: 0
+	}
 };
 
+Constants.WINDS = [Constants.EAST, Constants.SOUTH, Constants.WEST, Constants.NORTH];
```
Comment (Futochu) [inline imports/api/Constants.js]: Why is this defined outside the Constants object?

File: imports/api/NewGameUtils.js [hunk 11]
```diff
@@ -56,38 +56,36 @@ export var NewGameUtils = {
...
 		};
 	},
 
 	// Helper function to ensure all players are selected
 	allPlayersSelected() {
 		return (Session.get("current_east") != Constants.DEFAULT_EAST &&
-		        Session.get("current_south") != Constants.DEFAULT_SOUTH &&
-		        Session.get("current_west") != Constants.DEFAULT_WEST &&
-		        Session.get("current_north") != Constants.DEFAULT_NORTH);
+				Session.get("current_south") != Constants.DEFAULT_SOUTH &&
```
Comment (Futochu) [inline imports/api/NewGameUtils.js]: Align these blocks with spaces.  For instance, this line should be two tabs followed by 8 spaces in order for the Session(...) code to align properly on all environments.

File: imports/api/NewGameUtils.js [hunk 12]
```diff
@@ -129,13 +123,13 @@ export var NewGameUtils = {
 		let westRoundOver = Session.get("current_round") > 12;
 		// End condition where game has reached the end of south round with at least one player above minimum
 		let someoneAboveMinimum = Session.get("current_round") > 8 &&
-		                          this.someoneAboveMinimum(Constants.JPN_END_POINTS);
+								  this.someoneAboveMinimum(Constants.JPN_END_POINTS);
```
Comment (Futochu) [inline imports/api/NewGameUtils.js]: Same as above (RE: Spaces)

File: imports/api/NewGameUtils.js [hunk 13]
```diff
@@ -129,13 +123,13 @@ export var NewGameUtils = {
...
 		// End condition where north player reaches first place after winning on last round
 		let dealerFirstAndAboveMinimum = Session.get("current_round") == 8 &&
-		                                 Session.get("current_bonus") > 0 &&
-		                                 this.getDirectionScore("north") >= Constants.JPN_END_POINTS &&
-		                                 this.getFirstPlace() == "north";
+										 Session.get("current_bonus") > 0 &&
```
Comment (Futochu) [inline imports/api/NewGameUtils.js]: Same as above (RE: Spaces)

File: imports/api/PointCalculationUtils.js [hunk 14]
```diff
@@ -0,0 +1,106 @@
+import { Constants } from './Constants';
+import { NewGameUtils } from './NewGameUtils';
+
+export var PointCalculationUtils = {
+	jpn: {
+		dealin_delta,
```
Comment (Futochu) [inline imports/api/PointCalculationUtils.js]: Perhaps these should be renamed jpn_dealin_delta to more easily accomodate later addition of hkg

File: imports/api/PointCalculationUtils.js [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+				winds[w] = dealerPays + (nonDealerPays * 2) + (currentBonus * 300) + (riichiSticks * 1000);
+			} else {
+				winds[w] = -(nonDealerPays + currentBonus * 100);
+			}
+		} else {
+			if (w == winnerWind) {
+				winds[w] = (nonDealerPays * 3) + (currentBonus * 300) + (riichiSticks * 1000);
+			} else {
+				winds[w] = -(nonDealerPays + (currentBonus * 100));
+		}
```
Comment (TwelveNights) [inline imports/api/PointCalculationUtils.js]: Gotta correct this as an aside, lol

File: imports/api/PointCalculationUtils.js [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+
+export var PointCalculationUtils = {
+	jpn: {
+		dealin_delta,
+		selfdraw_delta,
+		mistake_delta
+	}
+};
+
+function dealin_delta(points, fu, winnerWind, loserWind, riichiSticks) {
```
Comment (Futochu) [inline imports/api/PointCalculationUtils.js]: Would it hurt to put comments in your code additions?  It makes it faster for others to review, and easier for others to maintain.  I think we should probably start including function headers as well just to make things easier to understand.

File: imports/api/PointCalculationUtils.js [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+		if (points == 1 && (fu == 20 || fu == 25)) {
+			return 0; // Issue Protection
+		} else {
+			basicPoints = fu * Math.pow(2, 2 + points);
+			basicPoints = basicPoints < Constants.JPN_MANGAN_BASE_POINTS ? basicPoints : Constants.JPN_MANGAN_BASE_POINTS;
+		}
+	} else { basicPoints = manganValue(points); }
+
+	let nonDealerPays = Math.ceil(basicPoints / 100 * (dealerWind == winnerWind ? 2 : 1)) * 100;
+	let dealerPays = Math.ceil(basicPoints / 100 * 2) * 100;
```
Comment (Futochu) [inline imports/api/PointCalculationUtils.js]: Should declare used variables at the beginning of the function.

File: imports/api/PointCalculationUtils.js [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+		if (fu == 20 || (points == 1 && fu == 25)) {
+			return 0; // Issue Protection
+		} else {
+			let manganPayout = Constants.JPN_MANGAN_BASE_POINTS * multiplier;
+			basicPoints = Math.ceil((fu * Math.pow(2, 2 + points)) * multiplier / 100) * 100;
+			basicPoints = basicPoints < manganPayout ? basicPoints : manganPayout;
+		}
+	} else { basicPoints = manganValue(points) * multiplier };
+
+	let currentBonus = Number(Session.get("current_bonus")) * 300;
```
Comment (Futochu) [inline imports/api/PointCalculationUtils.js]: Bonus amount should probably be defined in Constants.js

File: imports/api/PointCalculationUtils.js [hunk 19]
```diff
@@ -0,0 +1,106 @@
...
+	winds[winnerWind] = basicPoints + currentBonus + (riichiSticks * 1000);
```
Comment (Futochu) [inline imports/api/PointCalculationUtils.js]: So should riichi amount

File: imports/api/PointCalculationUtils.js [hunk 20]
```diff
@@ -0,0 +1,106 @@
...
+			}
+	}
+
+	Session.set("free_riichi_sticks", 0);
+	return winds;
+};
+
+function mistake_delta(loser) {
+	let winds = {};
+	winds[Constants.EAST] = winds[Constants.SOUTH] = winds[Constants.WEST] = winds[Constants.NORTH] = 4000;
```
Comment (Futochu) [inline imports/api/PointCalculationUtils.js]: Mistake points should probably be defined in Constants.js

File: imports/api/Constants.js (hunk 10 above)
Comment (TwelveNights) [inline imports/api/Constants.js]: Those values don't exist until the Constants object is defined. I added this line afterwards so I didn't have to be redundant.

File: imports/api/Constants.js [hunk 21]
```diff
@@ -13,11 +13,23 @@ export const Constants = {
...
+	JPN_MANGAN_BASE_POINTS: 2000,
+
+	// Points for every bonus round
+	BONUS_POINTS: 300,
+
+	// Points for every riichi stick
+	RIICHI_POINTS: 1000,
+
+	// Points paid for a mistake
+	MISTAKE_POINTS: 12000,
```
Comment (Futochu) [inline imports/api/Constants.js]: Please specify this is the Japanese chombo amount.  The Hong Kong amount is different.

Comment (Futochu) [inline imports/api/Constants.js]: Might as well give the JPN_ prefix to bonus and Riichi too, since that is sort of the convention it seems.

File: imports/api/NewGameUtils.js (hunk 11 above)
Comment (Futochu) [inline imports/api/NewGameUtils.js]: Did not address.

File: imports/api/NewGameUtils.js (hunk 12 above)
Comment (Futochu) [inline imports/api/NewGameUtils.js]: Did not address.

File: imports/api/NewGameUtils.js [hunk 22]
```diff
@@ -129,13 +123,12 @@ export var NewGameUtils = {
 		let westRoundOver = Session.get("current_round") > 12;
 		// End condition where game has reached the end of south round with at least one player above minimum
 		let someoneAboveMinimum = Session.get("current_round") > 8 &&
-		                          this.someoneAboveMinimum(Constants.JPN_END_POINTS);
+		                this.someoneAboveMinimum(Constants.JPN_END_POINTS);
```
Comment (Futochu) [inline imports/api/NewGameUtils.js]: So close, but this doesn't actually line up with the above line.

File: imports/api/NewGameUtils.js [hunk 23]
```diff
@@ -129,13 +123,12 @@ export var NewGameUtils = {
...
 		// End condition where north player reaches first place after winning on last round
 		let dealerFirstAndAboveMinimum = Session.get("current_round") == 8 &&
-		                                 Session.get("current_bonus") > 0 &&
-		                                 this.getDirectionScore("north") >= Constants.JPN_END_POINTS &&
-		                                 this.getFirstPlace() == "north";
-
+		                Session.get("current_bonus") > 0 &&
```
Comment (Futochu) [inline imports/api/NewGameUtils.js]: See above.

File: imports/api/PointCalculationUtils.js [hunk 24]
```diff
@@ -0,0 +1,132 @@
...
+	// Check to see if you have to count basic points
+	if (points < 5) {
+		if (fu == 20 || (points == 1 && fu == 25)) {
+			return 0; // Issue Protection
+		} else {
+			let manganPayout = Constants.JPN_MANGAN_BASE_POINTS * multiplier;
+			basicPoints = Math.ceil((fu * Math.pow(2, 2 + points)) * multiplier / 100) * 100;
+			basicPoints = basicPoints < manganPayout ? basicPoints : manganPayout;
+		}
+	} else { basicPoints = manganValue(points) * multiplier };
```
Comment (Futochu) [inline imports/api/PointCalculationUtils.js]: If you're going to do this, the semicolon should be at the end of the line, not the end of the block.  I would still make this and the other one like it a seperate line.

File: imports/api/PointCalculationUtils.js [hunk 25]
```diff
@@ -0,0 +1,132 @@
...
+	jpn_mistake_delta
+};
+
+/**
+ * Calculates the change in points as a result of a deal-in hand
+ * Uses the formula from https://en.wikipedia.org/wiki/Japanese_Mahjong_scoring_rules
+ * 
+ * @return {Object} containing the point difference for each seat
+ */
+function jpn_dealin_delta(points, fu, winnerWind, loserWind, riichiSticks) {
```
Comment (Futochu) [inline imports/api/PointCalculationUtils.js]: If we're adding function headers, make sure to add @param entries

File: imports/api/PointCalculationUtils.js [hunk 26]
```diff
@@ -0,0 +1,132 @@
+import { Constants } from './Constants';
+import { NewGameUtils } from './NewGameUtils';
+
+export var PointCalculationUtils = {
+	jpn_dealin_delta,
+	jpn_selfdraw_delta,
+	jpn_mistake_delta
```
Comment (Futochu) [inline imports/api/PointCalculationUtils.js]: Hmm, is it possible to keep the jpn object in PointCalculationUtils with members "dealin_delta = jpn_dealin_delta"?  It seems you removed it this entry in favour of just the suffix.

File: imports/ui/JapaneseNewGame.js [hunk 27]
```diff
@@ -801,14 +802,13 @@ function push_dealin_hand(template) {
...
-	eastDelta  += dealin_delta(points, fu, "east", winnerWind, loserWind)  +
-				  rewardRiichiSticks(riichiSum, "east", winnerWind);
-	southDelta += dealin_delta(points, fu, "south", winnerWind, loserWind) +
-				  rewardRiichiSticks(riichiSum, "south", winnerWind);
-	westDelta  += dealin_delta(points, fu, "west", winnerWind, loserWind)  +
-				  rewardRiichiSticks(riichiSum, "west", winnerWind);
-	northDelta += dealin_delta(points, fu, "north", winnerWind, loserWind) +
-				  rewardRiichiSticks(riichiSum, "north", winnerWind);
+	allDelta = PointCalculationUtils.jpn_dealin_delta(points, fu, winnerWind, loserWind, riichiSum);
+	Session.set("free_riichi_sticks", 0);
```
Comment (Futochu) [inline imports/ui/JapaneseNewGame.js]: 1. This is redundant as it exists in PointCalculationUtils (I would remove it from there personally)
2. This is located in a different location than in self_draw.  Should be consistent.

File: imports/api/PointCalculationUtils.js [hunk 28]
```diff
@@ -0,0 +1,147 @@
...
+		selfdraw_delta: jpn_selfdraw_delta,
+		mistake_delta: jpn_mistake_delta
+	}
+};
+
+/**
+ * Calculates the change in points as a result of a deal-in hand
+ * Uses the formula from https://en.wikipedia.org/wiki/Japanese_Mahjong_scoring_rules
+ *
+ * @param {number} points - the total number of han + dora
```
Comment (Futochu) [inline imports/api/PointCalculationUtils.js]: I think param variable types need to be capitalized.

File: imports/ui/JapaneseNewGame.js [hunk 29]
```diff
@@ -1089,36 +1091,37 @@ function push_split_pao_hand(template) {
 		Session.set("north_riichi_sum", Number(Session.get("north_riichi_sum")) + 1);
 	}
 
-	var value = dealin_delta(points, fu, winnerWind, winnerWind);
+	var value = PointCalculationUtils.jpn.dealin_delta(points, fu, winnerWind, loserWind, 0)[loserWind];
```
Comment (Futochu) [inline imports/ui/JapaneseNewGame.js]: Is this correct?  I'm worried this might not be fully tested.

File: imports/api/PointCalculationUtils.js [hunk 30]
```diff
@@ -0,0 +1,147 @@
...
+	let dealerPays;
+
+	let currentBonus = Number(Session.get("currentBonus"));
+	let individualBonusPayout = Constants.JPN_BONUS_POINTS / 3;
+
+	winds[Constants.EAST] = winds[Constants.SOUTH] = winds[Constants.WEST] = winds[Constants.NORTH] = 0;
+
+	// Check to see if you have to count basic points
+	if (points < 5) {
+		if (points == 1 && (fu == 20 || fu == 25)) {
```
Comment (Futochu) [inline imports/api/PointCalculationUtils.js]: So will this be removed or augmented with 2p25f?  In between makes no sense

File: imports/api/Constants.js [hunk 31]
```diff
@@ -43,7 +55,21 @@ export const Constants = {
 	SELF_DRAW: "selfdraw",
 	NO_WIN: "nowin",
 	RESTART: "restart",
-	MISTAKE: "fuckup"
+	MISTAKE: "fuckup",
```
Comment (Futochu) [inline imports/api/Constants.js]: This might conflict now

File: imports/ui/JapaneseNewGame.js [hunk 32]
```diff
@@ -790,7 +791,7 @@ function save_game_to_database(hands_array) {
 		if (Number(Session.get("west_score")) >= Number(Session.get("north_score"))) position--;
 		Players.update({_id: west_id}, {$inc: {japanesePositionSum: position}});
 
-		switch (position) {
+    switch (position) {
```
Comment (Futochu) [inline imports/ui/JapaneseNewGame.js]: There's generally a lot of ruined indentation in this patch.  Please fix.  (Tabs for indentation, spaces for alignment)

File: imports/ui/JapaneseNewGame.js [hunk 33]
```diff
@@ -1188,36 +1195,37 @@ function push_split_pao_hand(template) {
 		Session.set("north_riichi_sum", Number(Session.get("north_riichi_sum")) + 1);
 	}
 
-	var value = dealin_delta(points, fu, winnerWind, winnerWind);
+	var value = PointCalculationUtils.jpn.dealin_delta(points, fu, winnerWind, loserWind, 0)[winnerWind];
```
Comment (Futochu) [inline imports/ui/JapaneseNewGame.js]: It's worth creating another PointCalculationUtils function for Pao situations.

**Comment by Futochu** [comment]
Can you rebase so I can review this again?

**Comment by 0spooky** [comment]
Now a duplicate of #64
